### PR TITLE
Two-factor authentication feature in API part.

### PR DESCRIPTION
**Pull request by ecnelises**
This PR is for the multifactor auth of the API part. The changes are added into v1 version of rubygems.org API. Current design is:

* A new interface will be added into `ProfilesController` (or maybe other), getting the mfa setting status of current user (needs authentication).
* If the mfa on write option (_mfa\_login\_and\_write_) is enabled, those actions needing mfa should check `OTP` field in request header, and verify it.
* If verification on OTP fails, a 401 forbidden response with prompt text are returned. Otherwise, everything goes normally.
* Actions needing mfa are:
  * Pushing gems
  * Change ownership of gems.
* Maybe a new interface will be added for changing mfa status through RESTful API.

I'm not sure whether the approach is compatible for the rubygems versions (not supporting mfa). I will check how rubygems client handles the operation errors.

Below are tasks.
- [x] Add check for `OTP` field in header.
- [x] Check for rubygems push action.
- [x] Check for ownership change action.
- [ ] Check for creating deletions (`yank` command).
- [x] MFA level change. (In UI part)
- [x] Check for API Key acquirement.
- [x] Integration testing.
- [ ] (Maybe) Support for mfa state change (enabling, disabling, etc.) through API.

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -53,6 +53,13 @@ def redirect_to_root
     redirect_to root_path
   end
 
+  def verify_with_otp
+    return unless @api_user.mfa_login_and_write?
+    otp = request.headers["HTTP_OTP"] || ''
+    return if @api_user&.otp_verified?(otp)
```
Comment (sonalkr132) [inline app/controllers/application_controller.rb]: is this `&` needed here? `verify_authenticated_user` ensures that user see 401 if `@api_user` is nil.

**Comment by sonalkr132** [comment]
This looks good so far :+1:  Can we please make some headway with
> MFA level change. (In UI part)

**Comment by ecnelises** [comment]
Something like this seems reasonable. @sonalkr132 
change-mfa-level-plan

**Comment by sonalkr132** [comment]
> Something like this seems reasonable

Its fine but can be better.
Check these out:
https://codepen.io/cssinate/pen/KVdYjz
https://codepen.io/tr13ze/pen/yEdxvN
https://codepen.io/tag/dropdown/

[Public pens are MIT licensed](https://blog.codepen.io/legal/licensing/), if you are going to copy any of them, please make sure it matches our color scheme.

**Comment by ecnelises** [comment]
@sonalkr132 
Thanks, I'll check them out. I saw styles about `select` tags in our css file, but there's no `select` in the templates before.

File: app/models/user.rb [hunk 2]
```diff
@@ -193,6 +193,11 @@ def enable_mfa!(seed, level)
     save!(validate: false)
   end
 
+  def mfa_write_authorized?(otp)
+    return true unless mfa_login_and_write?
+    otp_verified?(otp || '')
```
Comment (sonalkr132) [inline app/models/user.rb]: please don't do `otp || ''`, either add default argument to `mfa_write_authorized?` or `otp_verified?`.

Comment (ecnelises) [reply app/models/user.rb]: I think I can replace `otp || ''` with `otp.to_s`

File: app/views/profiles/edit.html.erb [hunk 3]
```diff
@@ -90,9 +90,17 @@
...
       <p><%= t '.mfa.enabled' %></p>
-      <%= form_tag multifactor_auth_path, :method => :delete do %>
-        <%= text_field_tag :otp, '', :class => 'form__input' %>
-        <%= submit_tag t('.mfa.disable'), :class => 'form__submit' %>
+      <%= form_tag multifactor_auth_path, :method => :put, :id => 'mfa-edit' do %>
+        <%= label_tag :level, 'Auth Level', :class => 'form__label' %>
+        <%= select_tag :level, options_for_select([
+          ['Login and Write', 'mfa_login_and_write'],
+          ['Login Only', 'mfa_login_only'],
+          ['Disabled', 'no_mfa']], @user.mfa_level), :class => 'form__input form__select' %>
```
Comment (sonalkr132) [inline app/views/profiles/edit.html.erb]: `Login only` and `Login and write` don't make much sense to me. At least to user, we should show something more appropriate like `UI Authentication` and `UI and API Authentication`.

File: test/integration/profile_test.rb [hunk 4]
```diff
@@ -201,7 +203,8 @@ def enable_mfa
     recoveries = page.find_by_id("recovery-code-list").text.split
     click_link "Continue"
     page.fill_in "otp", with: recoveries.sample
-    click_button "Disable"
+    page.select "Disabled"
+    find('#mfa-edit input[type=submit]').click
```
Comment (sonalkr132) [inline test/integration/profile_test.rb]: Please make a function out of above two line.

File: test/functional/multifactor_auths_controller_test.rb [hunk 5]
```diff
@@ -34,42 +34,69 @@ class MultifactorAuthsControllerTest < ActionController::TestCase
...
+            should 'disable mfa' do
+              refute @user.reload.mfa_enabled?
+            end
           end
 
-          should respond_with :redirect
-          should redirect_to('the profile edit page') { edit_profile_path }
-          should 'disable mfa' do
-            refute @user.reload.mfa_enabled?
+          context 'when input recovery code' do
```
Comment (sonalkr132) [inline test/functional/multifactor_auths_controller_test.rb]: `when otp is recovery code`

File: app/views/profiles/edit.html.erb (hunk 3 above)
Comment (sonalkr132) [reply app/views/profiles/edit.html.erb]: We should also consider updating function names as well. I am not a stickler for naming it `UI and API Authentication` but `..login_and_write` is just misleading. For example, api request to show key is not a write operation and  by `login_only` we don't mean api login. Distinction has more to do with whether we are checking something for api or ui. 
Sorry, I failed to point it out previously.

File: config/locales/en.yml [hunk 6]
```diff
@@ -10,6 +10,7 @@ en:
   none: None
   not_found: Not Found
   please_sign_up: Access Denied. Please sign up for an account at https://rubygems.org
+  please_send_correct_otp: You have enabled multifactor authentication but your request doesn't have the correct OTP code. Please check it and retry.
```
Comment (sonalkr132) [inline config/locales/en.yml]: > your request doesn't have the correct OTP code.

Won't we be showing this when OPT is missing too? if yes, please update it to say `you didn't provide otp or provided otp was incorrect.`

**Comment by ecnelises** [comment]
The name _login and write_ or _login only_ derives from the original RFC issue about multi-factor authentication. NPM uses these names to identify their different authentication level. These are basic elements in mfa implementation, so I just borrowed them at that time without too much hesitation.

Currently, fetching API key requires OTP only when auth level set to _login and write_. Truly, these would be confusing to both users and contributors who see this code. Some of other actions not wrapped with mfa now, like API key reset or reset password, but they are also critical. Should they require OTP if _UI Authentication_ enabled? Except that, I think name around _UI_ and _API_ is good. Thanks.

File: app/views/profiles/edit.html.erb (hunk 3 above)
Comment (ecnelises) [reply app/views/profiles/edit.html.erb]: The name _login and write_ or _login only_ derives from the original RFC issue about multi-factor authentication. NPM uses these names to identify their different authentication level. These are basic elements in mfa implementation, so I just borrowed them at that time without too much hesitation.

Currently, fetching API key requires OTP only when auth level set to _login and write_. Truly, these would be confusing to both users and contributors who see this code. Some of other actions not wrapped with mfa now, like API key reset or reset password, but they are also critical. Should they require OTP if _UI Authentication_ enabled? Except that, I think name around _UI_ and _API_ is good. Thanks.

**Comment by sonalkr132** [comment]
> Some of other actions not wrapped with mfa now, like API key reset or reset password, but they are also critical.

In my opinion, asking mfa once during login through UI is good enough. I wish we could do the same for api as well, as pointed out by @colby-swandale 
> Can we just ask the user to give their MFA token whenever they sign in to an account which has 2FA?

We are asking for otp for every critical cli components because as of now, there is no concept of session when it comes to our api. Unless we implement a way of api authentication where tokens expires after a given time, we can't really do the same for api requests.
aws cli seem to be using related workflow https://aws.amazon.com/premiumsupport/knowledge-center/authenticate-mfa-cli/

**Comment by ecnelises** [comment]
> In my opinion, asking mfa once during login through UI is good enough.

Yes, but change mfa level still requires otp check. Actually we don’t have concept like _no mfa check after a duration_. In GitHub, if I’m going to delete one of my repo, the site would ask me to re-input my password. If I want to delete others, password is not needed in the short period. We don’t have this feature.

> I wish we could do the same for api as well.

I think it’s feasible to use JWT-like approach to implement it - wrap API key and _session duration_ with sign in header. There was talk about this before. I can handle it as further work.

Authentication of API just check header fields from every request. In ideal implementation, UI actions needing mfa can also be wrapped just with something like `before_action :mfa_ui_auth`, similar to how API does. User requests these actions, they will be redirected to a mfa prompt page, enter correct otp, finish actions, otherwise redirected back with a flash message.

File: app/views/profiles/edit.html.erb (hunk 3 above)
Comment (ecnelises) [reply app/views/profiles/edit.html.erb]: I did it. `Auth for UI Only` and `Auth for UI and API` is okay.

**Comment by hsbt** [comment]
ref https://github.com/rubygems/rubygems.org/issues/1725

File: app/controllers/api/v1/multifactor_auths_controller.rb [hunk 7]
```diff
@@ -0,0 +1,12 @@
+class Api::V1::MultifactorAuthsController < Api::BaseController
```
Comment (sonalkr132) [inline app/controllers/api/v1/multifactor_auths_controller.rb]: I don't think we are using this in client any more.

File: config/locales/en.yml [hunk 8]
```diff
@@ -10,6 +10,7 @@ en:
   none: None
   not_found: Not Found
   please_sign_up: Access Denied. Please sign up for an account at https://rubygems.org
+  otp_incorrect_or_missing: You have enabled multifactor authentication but your OTP code is missing or incorrect. Please check it and retry.
```
Comment (sonalkr132) [inline config/locales/en.yml]: > your OTP code is missing or incorrect.

this is unnecessarily ambiguous. Can we please add a check for otp being empty and show error message accordingly?

File: config/locales/en.yml [hunk 9]
```diff
@@ -158,6 +159,8 @@ en:
       note: "You MUST keep the recovery codes safe to prevent losing your account. Each of the code can be used once if you lose your authenticator."
     destroy:
       success: You have successfully disabled multifactor authentication.
+    update:
+      success: You have successfully updated your auth level.
```
Comment (sonalkr132) [inline config/locales/en.yml]: aren't we updating mfa level and not auth level?

File: config/locales/en.yml [hunk 10]
```diff
@@ -184,8 +187,8 @@ en:
         multifactor_auth: Multifactor authentication
         disabled: You have not yet enabled multifactor authentication.
         go_settings: Register a new device
-        enabled: You have enabled multifactor authentication. Please input your OTP from your authenticator or one of your active recovery codes to disable it.
-        disable: Disable
+        enabled: You have enabled multifactor authentication. Please input your OTP from your authenticator or one of your active recovery codes to change your auth level or disable it.
```
Comment (sonalkr132) [inline config/locales/en.yml]: mfa level

File: test/functional/api/v1/owners_controller_test.rb [hunk 11]
```diff
@@ -84,14 +84,57 @@ def self.should_respond_to(format)
...
 
-    should "add other user as gem owner with email" do
-      post :create, params: { rubygem_id: @rubygem.to_param, email: @second_user.email }, format: :json
-      assert @rubygem.owners.include?(@second_user)
+    context "when mfa for UI and API is enabled" do
+      setup do
+        @user.enable_mfa!(ROTP::Base32.random_base32, :ui_and_api_mfa)
+      end
+
+      context "on POST to add other user as gem owner with email without OTP" do
```
Comment (sonalkr132) [inline test/functional/api/v1/owners_controller_test.rb]: please drop `with email` in all these context.

**Comment by sonalkr132** [comment]
Rebase :collision: please.

File: config/locales/en.yml (hunk 8 above)
Comment (ecnelises) [reply config/locales/en.yml]: Right. But change to this may affect rubygems client, I'll check more of it.

**Comment by ecnelises** [comment]
Done.

File: app/controllers/application_controller.rb [hunk 12]
```diff
@@ -53,6 +53,13 @@ def redirect_to_root
     redirect_to root_path
   end
 
+  def verify_with_otp
+    otp = request.headers["HTTP_OTP"]
+    return if @api_user.mfa_api_authorized?(otp)
+    prompt_text = otp.present? ? t(:otp_incorrect) : t(:otp_missing)
```
Comment (sonalkr132) [inline app/controllers/application_controller.rb]: nice :100:

File: app/views/profiles/edit.html.erb [hunk 13]
```diff
@@ -90,9 +90,17 @@
     <h2><%= t '.mfa.multifactor_auth' %></h2>
     <% if @user.mfa_enabled? %>
       <p><%= t '.mfa.enabled' %></p>
-      <%= form_tag multifactor_auth_path, :method => :delete do %>
-        <%= text_field_tag :otp, '', :class => 'form__input' %>
-        <%= submit_tag t('.mfa.disable'), :class => 'form__submit' %>
+      <%= form_tag multifactor_auth_path, :method => :put, :id => 'mfa-edit' do %>
+        <%= label_tag :level, 'MFA Level', :class => 'form__label' %>
+        <%= select_tag :level, options_for_select([
+          ['Auth for UI and API', 'ui_and_api_mfa'],
```
Comment (sonalkr132) [inline app/views/profiles/edit.html.erb]: hi, we need to move these to locales.

File: app/controllers/api/v1/api_keys_controller.rb [hunk 14]
```diff
@@ -5,12 +5,16 @@ class Api::V1::ApiKeysController < Api::BaseController
   def show
     authenticate_or_request_with_http_basic do |username, password|
       user = User.authenticate(username, password)
-      if user
+      otp = request.headers['HTTP_OTP']
+      if user&.mfa_api_authorized?(otp)
```
Comment (sonalkr132) [inline app/controllers/api/v1/api_keys_controller.rb]: why aren't we using `before_action :verify_with_otp` here?

Comment (ecnelises) [reply app/controllers/api/v1/api_keys_controller.rb]: `verify_with_otp` requires `@api_user` set. We only require username & password in `ApiKeys#show` and `Sessions#create`.

File: config/locales/en.yml [hunk 15]
```diff
@@ -184,8 +188,13 @@ en:
...
         go_settings: Register a new device
-        enabled: You have enabled multifactor authentication. Please input your OTP from your authenticator or one of your active recovery codes to disable it.
-        disable: Disable
+        enabled: You have enabled multifactor authentication. Please input your OTP from your authenticator or one of your active recovery codes to change your MFA level or disable it.
+        update: Update
+        level:
+          title: MFA Level
+          no_mfa: Disabled
+          ui_mfa_only: Auth for UI Only
+          ui_and_api_mfa: Auth for UI and API
```
Comment (sonalkr132) [inline config/locales/en.yml]: Please remove `Auth for`. UI Only and, UI and API will do.

**Comment by sonalkr132** [comment]
One minor change. Looks good to merge now :+1: 
Have you tried this locally with https://github.com/rubygems/rubygems/pull/2369?

**Comment by ecnelises** [comment]
> One minor change. Looks good to merge now 👍
> Have you tried this locally with [rubygems/rubygems#2369](https://github.com/rubygems/rubygems/pull/2369)?

Local output is like this:
https://paste.ofcode.org/h473UpYsJGytsarQnxjbWC

One thing I need to do: I have to comment the code checking current RubyGems client version. Or I would get

> You are using a beta release of RubyGems (3.0.0.beta1) which is not
allowed to push gems.  Please downgrade or upgrade to a release version.

**Comment by sonalkr132** [comment]
> This command needs digit code for multifactor authentication.
Code:   757891
Signed in.
Pushing gem to http://localhost:3000...
This command needs digit code for multifactor authentication.
Code:   757891
Successfully registered gem: hello (0.0.9)

If we don't delete `otp` [here](https://github.com/rubygems/rubygems/pull/2369/files#diff-8c06a0bbeb02bbae5583a453bc7b16dfR135), it would be nicer not to ask for otp again.

**Comment by ecnelises** [comment]
> If we don't delete otp here, it would be nicer not to ask for otp again.

I remember we talked about it and I checked the chat history. It seems that I misunderstood your meanings. Currently, I need to do multi-factor authentication even seconds after signing in with correct OTP code. Security problem arises when such no-otp interval exists. But now we don't have it. Every execution is independent.

Only concern: If fetching newest version of `gem` is too slow, the OTP may expire. (Since we have tolerance for OTP in last interval, this should not be a big problem)

Thank you for notifying me about this. I'll delete first it until we found [new approaches for authentication](https://github.com/rubygems/rubygems.org/pull/1753#issuecomment-412322868).

**Comment by sonalkr132** [comment]
giphy-downsized-large

Thanks @ecnelises :wine_glass: :sparkles:

**Pull request merged**